### PR TITLE
More AppContext cleanups

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Remove getters that mutate the context
- [x] Add utility method `putIfAbsent` to context
- [x] Explicitly initialize the context in the executable entry point and the test context.

@devoncarew @tvolkert

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@pylaligand FYI

**Comment by pylaligand** [comment]
LGTM

File: packages/flutter_tools/lib/src/base/context.dart [hunk 1]
```diff
@@ -38,6 +38,16 @@ class AppContext {
 
   void operator[]=(Type type, dynamic instance) => setVariable(type, instance);
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/context.dart]: Can we nuke this now?

File: packages/flutter_tools/lib/src/toolchain.dart [hunk 2]
```diff
@@ -27,16 +27,13 @@ const Map<HostTool, String> _kHostToolFileName = const <HostTool, String>{
 class ToolConfiguration {
   /// [overrideCache] is configurable for testing.
   ToolConfiguration({ Cache overrideCache }) {
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/toolchain.dart]: Why not have the tests just use context.putIfAbsent()?

File: packages/flutter_tools/test/src/context.dart [hunk 3]
```diff
@@ -32,49 +34,34 @@ void testUsingContext(String description, dynamic testMethod(), {
   test(description, () async {
     AppContext testContext = new AppContext();
 
+    // Apply all overrides to the test context.
     overrides.forEach((Type type, dynamic value) {
       testContext[type] = value;
```
Comment (tvolkert) [inline packages/flutter_tools/test/src/context.dart]: Just use putIfAbsent() so we can remove the direct mutator?

File: packages/flutter_tools/lib/src/base/context.dart (hunk 1 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/base/context.dart]: Done.

File: packages/flutter_tools/lib/src/toolchain.dart (hunk 2 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/toolchain.dart]: Done. Much nicer!

File: packages/flutter_tools/test/src/context.dart (hunk 3 above)
Comment (johnmccutchan) [inline packages/flutter_tools/test/src/context.dart]: I'm removing the `[]=` operator but there are still places that we want to override something in a context so I'm leaving `setVariable`

**Comment by tvolkert** [comment]
lgtm!

**Pull request merged**